lib/puppet/provider/cisco_itd_service/cisco.rb

```diff
@@ -0,0 +1,256 @@
+  # after all the other properties are set. For all other cases,
+  # shutdown needs to be set first before any other
+  # properties are set. Basically, no properties cannot be
+  # changed while the service is active.
+  ITDSERVICE_SHUT_PROP = [
+    :shutdown
+  ]
+  ITDSERVICE_ARRAY_FLAT_PROPS = [
+    :ingress_interface,
+    :peer_vdc,
+    :virtual_ip,
+  ]
+  ITDSERVICE_ALL_PROPS = ITDSERVICE_NON_BOOL_PROPS +
+                         ITDSERVICE_ARRAY_FLAT_PROPS + ITDSERVICE_BOOL_PROPS
+  ITDSERVICE_ALL_BOOL_PROPS = ITDSERVICE_BOOL_PROPS + ITDSERVICE_SHUT_PROP
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@nu',
+                                            ITDSERVICE_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:array_flat, self, '@nu',
+                                            ITDSERVICE_ARRAY_FLAT_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@nu',
+                                            ITDSERVICE_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@nu',
+                                            ITDSERVICE_SHUT_PROP)
+
+  def initialize(value={})
+    super(value)
+    @nu = Cisco::ItdService.itds[@property_hash[:name]]
+    @property_flush = {}
+  end
+
+  def self.properties_get(itd_service_name, nu_obj)
+    debug "Checking instance, #{itd_service_name}."
+    current_state = {
+      service_name: itd_service_name,
+      name:         itd_service_name,
+      ensure:       :present,
+    }
+    # Call node_utils getter for each property
+    ITDSERVICE_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+    ITDSERVICE_ARRAY_FLAT_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+    ITDSERVICE_ALL_BOOL_PROPS.each do |prop|
+      val = nu_obj.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    # nested array properties
+    current_state[:ingress_interface] = nu_obj.ingress_interface
+    current_state[:peer_vdc] = nu_obj.peer_vdc
+    current_state[:virtual_ip] = nu_obj.virtual_ip
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    itds = []
+    Cisco::ItdService.itds.each do |itd_service_name, nu_obj|
+      itds << properties_get(itd_service_name, nu_obj)
+    end
+    itds
+  end
+
+  def self.prefetch(resources)
+    itds = instances
+    resources.keys.each do |name|
+      provider = itds.find { |nu_obj| nu_obj.instance_name == name }
+      resources[name].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def exists?
+    (@property_hash[:ensure] == :present)
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def instance_name
+    service_name
+  end
+
+  def shutdown_set
```

Comment:
@saichint This method and use of this method is a bit confusing.  Message me when you get in today so we can chat.